File: Python/reshape-the-matrix.py
# Time:  O(m * n)
# Space: O(m * n)

class Solution(object):
    def matrixReshape(self, nums, r, c):
        """
        :type nums: List[List[int]]
        :type r: int
        :type c: int
        :rtype: List[List[int]]
        """
        if not nums or \
           r*c != len(nums) * len(nums[0]):
            return nums

        result = [[0 for _ in xrange(c)] for _ in xrange(r)]
        count = 0
        for i in xrange(len(nums)):
            for j in xrange(len(nums[0])):
                result[count/c][count%c] = nums[i][j]
                count += 1
        return result

File: C++/reshape-the-matrix.cpp
// Time:  O(m * n)
// Space: O(m * n)

class Solution {
public:
    vector<vector<int>> matrixReshape(vector<vector<int>>& nums, int r, int c) {
        if (nums.empty() ||
            r * c != nums.size() * nums[0].size()) {
            return nums;
        }

        vector<vector<int>> result(r, vector<int>(c));
        int count = 0;
        for (int i = 0; i < nums.size(); ++i) {
            for (int j = 0; j < nums[0].size(); ++j) {
                result[count / c][count % c] = nums[i][j];
                ++count;
            }
        }
        return result;
    }
};
